federation/cluster/federation-up.sh

```diff
@@ -18,15 +18,66 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-KUBE_ROOT=$(readlink -m $(dirname "${BASH_SOURCE}")/../../)
+# This script is only used for e2e tests! Don't use it in production!
+# This is also a temporary bridge to slowly switch over everything to
+# federation/develop.sh. Carefully moving things step-by-step, ensuring
+# things don't break.
+# TODO(madhusudancs): Remove this script and its dependencies.
 
-. ${KUBE_ROOT}/federation/cluster/common.sh
 
-tagfile="${KUBE_ROOT}/federation/manifests/federated-image.tag"
-if [[ ! -f "$tagfile" ]]; then
-    echo "FATAL: tagfile ${tagfile} does not exist. Make sure that you have run build-tools/push-federation-images.sh"
-    exit 1
-fi
-export FEDERATION_IMAGE_TAG="$(cat "${KUBE_ROOT}/federation/manifests/federated-image.tag")"
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/../..
+# For `kube::log::status` function since it already sources
+# "${KUBE_ROOT}/cluster/lib/logging.sh" and DEFAULT_KUBECONFIG
+source "${KUBE_ROOT}/cluster/common.sh"
+# For $KUBE_PLATFORM, $KUBE_ARCH, $KUBE_BUILD_STAGE,
+# $FEDERATION_PUSH_REPO_BASE and $FEDERATION_NAMESPACE.
+source "${KUBE_ROOT}/federation/cluster/common.sh"
+# For `get_version` function and $KUBE_REGISTRY.
+# TODO(madhusudancs): Remove this when the code here is moved
+# to federation/develop.sh
+source "${KUBE_ROOT}/federation/develop/develop.sh"
 
-create-federation-api-objects
+FEDERATION_NAME="${FEDERATION_NAME:-e2e-federation}"
+DNS_ZONE_NAME="${FEDERATION_DNS_ZONE_NAME:-}"
+HOST_CLUSTER_CONTEXT="${FEDERATION_HOST_CLUSTER_CONTEXT:-${1}}"
+
+readonly CLIENT_BIN_DIR="${KUBE_ROOT}/_output/${KUBE_BUILD_STAGE}/client/${KUBE_PLATFORM}-${KUBE_ARCH}/kubernetes/client/bin"
+kubefed="${CLIENT_BIN_DIR}/kubefed"
+kubectl="${CLIENT_BIN_DIR}/kubectl"
+
+# Initializes the control plane.
+# TODO(madhusudancs): Move this to federation/develop.sh.
+function init() {
+  kube::log::status "Deploying federation control plane for ${FEDERATION_NAME} in cluster ${HOST_CLUSTER_CONTEXT}"
+
+  local -r kube_version="$(get_version)"
+
+  ${kubefed} init \
+      "${FEDERATION_NAME}" \
+      --host-cluster-context="${HOST_CLUSTER_CONTEXT}" \
+      --dns-zone-name="${DNS_ZONE_NAME}" \
+      --image="${KUBE_REGISTRY}/hyperkube-amd64:${kube_version}"
+}
+
+# create_cluster_secrets creates the secrets containing the kubeconfigs
+# of the participating clusters in the host cluster. The kubeconfigs itself
+# are created while deploying clusters, i.e. when kube-up is run.
+function create_cluster_secrets() {
```

Comment:
I am not entirely sure if we want to do this in the framework or using `kubefed join`. Each e2e test should have the autonomy to make this decision. We have tests (and we want them) that test the API server without having the clusters registered. Pre-registering clusters will make it difficult to run these tests. They will have to deregister the pre-registered clusters before running their tests and then register those clusters back. It kind of reverses the setup and teardown order which will in turn make the tests hard to reason about and debug.

We also need tests that incrementally register/deregister clusters. For example, start with 1 cluster, add two, delete one and so on and test how the controllers behave. We don't have good coverage for these kinds of operations today. But the point is, pre-registering clusters makes it hard to write these kinds of tests.